File: leetcode/medium/UniquePaths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        """
        Calculates the number of unique paths a robot can take to get from the
        top-left corner to the bottom-right corner of an m x n grid.
        The robot can only move either down or right at any point in time.

        This is solved using dynamic programming with space optimization.
        """
        # The core idea is that the number of paths to reach cell (i, j) is the
        # sum of paths to reach (i-1, j) and paths to reach (i, j-1).
        #
        # To optimize space from O(m*n) to O(n), we only need to store the
        # results of the previous row to calculate the current row.
        
        # We use a 1D array `dp` of size `n`. Initialize it with all 1s.
        # This represents the base case for the first row, where there's only
        # one way to reach each cell (by moving right).
        dp = [1] * n
        
        # Iterate through the grid, starting from the second row (i=1).
        for i in range(1, m):
            # For each row, iterate through the columns, starting from the
            # second column (j=1).
            for j in range(1, n):
                # dp[j] is the cell above, dp[j-1] is the cell to the left.
                dp[j] = dp[j] + dp[j-1]
                
        # The last element of the `dp` array will hold the total number of paths.
        return dp[n-1]

File: leetcode/medium/UniquePaths.java
import java.util.Arrays;

/**
 * Solves the Unique Paths problem using Space-Optimized Dynamic Programming.
 */
class Solution {

    /**
     * Calculates the number of unique paths from the top-left to bottom-right corner.
     *
     * @param m The number of rows in the grid.
     * @param n The number of columns in the grid.
     * @return The total number of unique paths.
     */
    public int uniquePaths(int m, int n) {
        // Base case: If the grid is 1x1 or invalid, return appropriate value (usually 1 or 0)
        if (m <= 0 || n <= 0) return 0;
        if (m == 1 || n == 1) return 1;

        // Optimization: Ensure we always allocate the smaller dimension to save space.
        // If rows (m) are fewer than cols (n), we swap them.
        if (m < n) {
            int temp = m;
            m = n;
            n = temp;
        }

        // dp[j] represents the number of paths to reach the cell at current row and column j.
        int[] dp = new int[n];
        
        // Initialize the first row (or column technically, due to swap).
        // There is always exactly 1 way to reach any cell in the first row (move right only).
        Arrays.fill(dp, 1);

        // Iterate starting from the second row
        for (int i = 1; i < m; i++) {
            // Iterate through the columns
            for (int j = 1; j < n; j++) {
                // The number of paths to current cell (i, j) is the sum of:
                // 1. Paths from the cell above (currently stored in dp[j])
                // 2. Paths from the cell to the left (stored in dp[j-1])
                dp[j] = dp[j] + dp[j - 1];
            }
        }

        // The last element contains the paths to the bottom-right corner
        return dp[n - 1];
    }
}